### PR TITLE
also migrate python 3.12 for numpy 2.0; "soft-close" 3.12 migration

### PR DESCRIPTION
**Pull request by h-vetinari**
Otherwise we'll need another migration just for 3.12, which is even more of a hassle than dealing with superfluous migrator PRs now. For more details/discussion, see https://github.com/conda-forge/conda-forge.github.io/issues/1997

### CODE REVIEW

**Comment by xhochy** [comment]
To understand correctly: This is now a combined migration that would pull a feedstock onto NumPy 2 and add Python 3.12 builds? 

OK with me, just want to make sure I understand it.

**Comment by beckermr** [comment]
Agreed this is fine but also maybe we just close the python 3.12 migration? That will have the same effect for PRs in the numpy2 migration and it is May anyways so why not?

**Review by hmaarrfk — CHANGES_REQUESTED** [review]
I would like to wait until a few numpy core maintainers have finished their disucssion as to whether or not python 3.9 should be even be included.

https://mail.python.org/archives/list/numpy-discussion@python.org/thread/AHTATJKGUEOILBNUI5IGGZPXJ5FXIRAU/

File: recipe/migrations/numpy2.yaml
Comment (h-vetinari) [reply recipe/migrations/numpy2.yaml:65]: I don't think this is relevant for conda-forge, where we can repopatch existing builds to have a `<2.0` constraint (see https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/712). So old scikit-image builds would never break by having py3.9 builds for numpy 2.0.

**Comment by h-vetinari** [comment]
> @beckermr: Agreed this is fine but also maybe we just close the python 3.12 migration? That will have the same effect for PRs in the numpy2 migration and it is May anyways so why not?

We still have around ~900 unmigrated feedstocks for 3.12 (see [comment](https://github.com/conda-forge/conda-forge.github.io/issues/1997#issuecomment-2094977305)). I think this is probably too much...? Also, in recent years we've closed the 3.X migration once we started migrating for 3.{X+1}.

> @xhochy: To understand correctly: This is now a combined migration that would pull a feedstock onto NumPy 2 and add Python 3.12 builds?

Yes exactly, and it overrides existing `python312.yaml` files as well (because "2.0" > "1.26").

The downside is that we'd be migrating all python-feedstocks (not just all numpy ones), see same [comment](https://github.com/conda-forge/conda-forge.github.io/issues/1997#issuecomment-2094977305).

File: recipe/migrations/numpy2.yaml
Comment (hmaarrfk) [reply recipe/migrations/numpy2.yaml:65]: lets follow numpy upstream for their supported workflows.

Ultimately, I want to stop us from "creating all the patches" for other projects to support numpy 2.0 + python 3.9.

Our job should "end" at one point in time.

Comment (hmaarrfk) [reply recipe/migrations/numpy2.yaml:65]: Ultimately, I want to respect and the to help the numpy developers in this transition.

If our goal is to proactively transition packages and report bugs upstream, we should do that, we shouldn't be creating more testing surface, and breakage, than is necessary.

I strongly believe that if numpy upstream developers decide to end support for an old version of python for numpy 2, we should honor that decision.

If they want to keep it, we should use our ability to help identify breakages ahead of a wider spread release.

Comment (h-vetinari) [reply recipe/migrations/numpy2.yaml:65]: > lets follow numpy upstream for their supported workflows.

Which you're proposing to change last minute. Of course, we should follow that for numpy 2.0 GA, but it doesn't have any bearing IMO on the rc-migration.

Comment (hmaarrfk) [reply recipe/migrations/numpy2.yaml:65]: Isn't the whole point of the rc-migration to stay ahead of the large re-compilation influx and to better predict breakages? 

Why would we increase the testing surface for many of the upstream developers (many numpy folk here and the rest of the open source science stack)?

Real concerns against widestream patching were brought up https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/712

We should be weary of wide scale patching and intrusive behavior. Again, I'm trying to honor upstream's wishes here, and I've made my point to them. We should wait a day or two for them to decide.

Comment (h-vetinari) [reply recipe/migrations/numpy2.yaml:65]: > Isn't the whole point of the rc-migration to stay ahead of the large re-compilation influx and to better predict breakages?

I'd say "prepare as many feedstocks as possible for a post-2.0 world before GA", but that's pretty much the same.

> Why would we increase the testing surface for many of the upstream developers?

I think that's red herring. In 99% of the cases, a package will either build or not build with numpy 2.0. For projects still supporting 3.9, I strongly doubt that this will create additional failure modes compared to those existing for 3.10+.

Also, this only concerns the rc-migration; I've already said we should match what numpy 2.0 GA does when we start publishing to the main channel (e.g. we could simply never publish the 3.9 build of numpy 2.0 there).

> Real concerns against widestream patching were brought up [conda-forge/conda-forge-repodata-patches-feedstock#712](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/712)

Sure, and they can be addressed (which @beckermr offered to do).

> Again, I'm trying to honor upstream's wishes here

That's a bit of a stretch IMO. They're your wishes, which upstream might choose to adopt, AFAICT mainly because PyPI doesn't have a way to do retroactive constraint patches.

> We should wait a day or two for them to decide.

It's fine to wait a day or two, sure. What I don't want is to hold up the migration for concerns that really only matter for numpy 2.0 GA. If we build a couple extra builds during the rc-migration for 3.9, that's completely harmless (unless for builds failing specifically on 3.9, who can then skip that), but if we migrate without 3.9 we cannot go back without a lot of manual effort if upstream decides to keep 3.9 support after all.

Comment (hmaarrfk) [reply recipe/migrations/numpy2.yaml:65]: Thanks for waiting a day or two, i know it isn't desirable when things are so close after such a long time.

**Comment by xhochy** [comment]
> Also, in recent years we've closed the 3.X migration once we started migrating for 3.{X+1}.

That is also only a coincidence of someone spending time on the Python migrations in general. After 4-6 months most packages have been migrated to the new Python version or will never be migrated as they are no longer maintained.  @ocepaf or myself have gone through all open PRs in the last years and were able to migrate a lot of feedstocks were upstream was actively maintained and the feedstocks not. I think once someone has done some passes like that, we can close 3.12.

**Comment by h-vetinari** [comment]
> I think once someone has done some passes like that, we can close 3.12.

Well, in that sense, going with the PR as-is would give every feedback another shot at being migrated for python 3.12, and with the attention the numpy2 migration will get, we could then definitely close it.

But I'm also fine with closing the 312 migration beforehand to avoid the PR churn, if people prefer that overall.

**Review by hmaarrfk — APPROVED** [review]
My concerns are no longer. thank you for waiting.

**Comment by jakirkham** [comment]
Sorry if I'm missing something, but have we already tested these changes in a feedstock somewhere?

**Comment by h-vetinari** [comment]
Thanks @hmaarrfk.

@conda-forge/core - I'd like to have some explicit (dis)agreement here... Do we:
1. think we can handle the extra PR churn (it would be easy just to close PRs if the feedstock doesn't build against numpy)?
2. prefer to just close the python 3.12 migration and put the onus on feedstocks that haven't migrated py312 yet to skip it for the time being?

I think 2. is more friendly to feedstocks that are "good citizens" and kept up-to-date, whereas 1. gives more time for people to finish migrating for 3.12, in exchange for more migration PR churn.

**Comment by jakirkham** [comment]
So I tried out the NumPy 2 migrator on a feedstock that uses NumPy via Python only (no C API). Here is the PR: https://github.com/conda-forge/numcodecs-feedstock/pull/106

AFAICT the only change it makes is adding `conda-forge/label/numpy_rc` to `channel_sources`. Am guessing this is what we expect?

If so, then the only reason the migrator will open a migration PR to a feedstock not using NumPy (in `build` or at all) is this channel addition. So if we were able to handle this channel addition another way, the migration bot would actually skip feedstocks (like this one) where NumPy is used at runtime or not at all

So how might we do this? Some options:

1. Add the channel in [`conda-forge-ci-setup`]( https://github.com/conda-forge/conda-forge-ci-setup-feedstock )
2. Add the channel globally and constrain all `numpy` versions to `<2` (which the migrator then removes)
3. Add the channel in the minimigrator
4. Having the NumPy 2 migrator wait for the NumPy 2 minimigrator (we did this successfully with [the CUDA 11.2 arch migrator]( https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/57e963da62fffaf4d44bd3e4a068c34c5aa3533f/recipe/migrations/cuda_112_ppc64le_aarch64.yaml#L16-L17 ))
5. ...?

Please feel free to suggest more

Admittedly all of these have their tradeoffs. That said, maybe this is a way we could find the balance between having the Python 3.12 migrator stay open a bit while keeping the NumPy 2 migrator focused on relevant feedstocks

**Comment by h-vetinari** [comment]
Thanks for the testing and additional ideas!

> Some options:

If I understand you correctly, then 1. & 2. need https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/712, which has it's own set of problems (i.e. patching ~100k artefacts; though I'd be in favour of it...).

Not sure about 3. - the thing needs to be rerender-proof. How would you add it in the piggyback so that it persists (except literally writing `channel_sources:` to the local CBC, which sounds too intrusive to me, and also hard to undo later).

I think 4. is not feasible currently. There's a piggyback that attaches to _some_ migration (or even all of them, if you really want), but it's not a migrator in and of itself that can be awaited (unlike the alt-arch migration).

> If so, then the only reason the migrator will open a migration PR to a feedstock not using NumPy (in `build` or at all) is this channel addition.

To be fair, any feedstock that hasn't been rerendered in a few weeks will produce a non-empty diff upon rerendering, and so the bot will open a PR regardless, because it cannot (or at least does not) distinguish based on the kind of changes that the diff produces (c.f. https://github.com/regro/cf-scripts/issues/2500).

**Comment by h-vetinari** [comment]
> @conda-forge/core - I'd like to have some explicit (dis)agreement here...

While waiting for responses here, I've started looking at winding down the python 3.12 migration: https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/5892

**Comment by h-vetinari** [comment]
For now I've followed the good idea of @hmaarrfk in #5892 to "soft-close" the 3.12 migration, as in: apply the changes of the migrator to the global pinning already, but do not delete the migrator file.

This has the advantage that we can still follow the bot's progress on the status page and the bot will also keep raising PRs for feedstocks once all their dependencies become available (even though, strictly speaking, a rerender of said feedstock would already pull in 3.12; feedstocks with missing dependencies would then have to add `skip: true  # [py>311]` until the bot come by).

I've tested the state of this PR in https://github.com/conda-forge/pythran-feedstock/pull/85 (the global pinning cannot be faked by `use_local: true`, but it's possible to override it when using `conda smithy rerender -e <global_pinning>`), and it works fine.

**Comment by h-vetinari** [comment]
This was just discussed in the core call and received a lot of support and no opposition.

**Comment by jakirkham** [comment]
Thanks Axel! 🙏

**Pull request merged**
